File: src/main/java/dev/cosgy/TextToSpeak/gui/TextAreaOutputStream.java
//////////////////////////////////////////////////////////////////////////////////////////
//  Copyright 2023 Cosgy Dev                                                             /
//                                                                                       /
//     Licensed under the Apache License, Version 2.0 (the "License");                   /
//     you may not use this file except in compliance with the License.                  /
//     You may obtain a copy of the License at                                           /
//                                                                                       /
//        http://www.apache.org/licenses/LICENSE-2.0                                     /
//                                                                                       /
//     Unless required by applicable law or agreed to in writing, software               /
//     distributed under the License is distributed on an "AS IS" BASIS,                 /
//     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.          /
//     See the License for the specific language governing permissions and               /
//     limitations under the License.                                                    /
//////////////////////////////////////////////////////////////////////////////////////////

package dev.cosgy.TextToSpeak.gui;

import javax.swing.*;
import java.awt.*;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * @author Kosugi_kun
 */
public class TextAreaOutputStream extends OutputStream {
    private final byte[] oneByte;
    private Appender appender;

    public TextAreaOutputStream(JTextArea txtara) {
        this(txtara, 1000);
    }

    public TextAreaOutputStream(JTextArea txtara, int maxlin) {
        if (maxlin < 1) {
            throw new IllegalArgumentException("TextAreaOutputStreamの最大行数は正数でなければなりません(value=" + maxlin + ")");
        }
        oneByte = new byte[1];
        appender = new Appender(txtara, maxlin);
    }

    static private String bytesToString(byte[] ba, int str, int len) {
        try {
            return new String(ba, str, len, System.getProperty("file.encoding"));
        } catch (UnsupportedEncodingException thr) {
            return new String(ba, str, len);
        }
    }

    public synchronized void clear() {
        if (appender != null) {
            appender.clear();
        }
    }

    @Override
    public synchronized void close() {
        appender = null;
    }

    @Override
    public synchronized void flush() {
        /* empty */
    }

    @Override
    public synchronized void write(int val) {
        oneByte[0] = (byte) val;
        write(oneByte, 0, 1);
    }

    @Override
    public synchronized void write(byte[] ba) {
        write(ba, 0, ba.length);
    }

    @Override
    public synchronized void write(byte[] ba, int str, int len) {
        if (appender != null) {
            appender.append(bytesToString(ba, str, len));
        }
    }

    static class Appender
            implements Runnable {
        static private final String EOL1 = "\n";
        static private final String EOL2 = System.getProperty("line.separator", EOL1);

        private final JTextArea textArea;
        private final int maxLines;
        private final LinkedList<Integer> lengths;
        private final List<String> values;

        private int curLength;
        private boolean clear;
        private boolean queue;

        Appender(JTextArea txtara, int maxlin) {
            textArea = txtara;
            maxLines = maxlin;
            lengths = new LinkedList<>();
            values = new ArrayList<>();

            curLength = 0;
            clear = false;
            queue = true;
        }

        private synchronized void append(String val) {
            values.add(val);
            if (queue) {
                queue = false;
                EventQueue.invokeLater(this);
            }
        }

        private synchronized void clear() {
            clear = true;
            curLength = 0;
            lengths.clear();
            values.clear();
            if (queue) {
                queue = false;
                EventQueue.invokeLater(this);
            }
        }

        @Override
        public synchronized void run() {
            if (clear) {
                textArea.setText("");
            }
            values.stream()
                    .peek((val) -> curLength += val.length())
                    .peek((val) -> {
                        if (val.endsWith(EOL1) || val.endsWith(EOL2)) {
                            if (lengths.size() >= maxLines) {
                                textArea.replaceRange("", 0, lengths.removeFirst());
                            }
                            lengths.addLast(curLength);
                            curLength = 0;
                        }
                    }).forEach(textArea::append);
            values.clear();
            clear = false;
            queue = true;
        }
    }
}

File: src/main/java/dev/cosgy/textToSpeak/gui/TextAreaOutputStream.kt
//////////////////////////////////////////////////////////////////////////////////////////
//  Copyright 2023 Cosgy Dev                                                             /
//                                                                                       /
//     Licensed under the Apache License, Version 2.0 (the "License");                   /
//     you may not use this file except in compliance with the License.                  /
//     You may obtain a copy of the License at                                           /
//                                                                                       /
//        http://www.apache.org/licenses/LICENSE-2.0                                     /
//                                                                                       /
//     Unless required by applicable law or agreed to in writing, software               /
//     distributed under the License is distributed on an "AS IS" BASIS,                 /
//     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.          /
//     See the License for the specific language governing permissions and               /
//     limitations under the License.                                                    /
//////////////////////////////////////////////////////////////////////////////////////////
package dev.cosgy.textToSpeak.gui

import java.awt.EventQueue
import java.io.OutputStream
import java.io.UnsupportedEncodingException
import java.nio.charset.Charset
import java.util.*
import javax.swing.JTextArea

/**
 * @author Kosugi_kun
 */
class TextAreaOutputStream @JvmOverloads constructor(txtara: JTextArea, maxlin: Int = 1000) : OutputStream() {
    private val oneByte: ByteArray
    private var appender: Appender?

    init {
        require(maxlin >= 1) { "TextAreaOutputStreamの最大行数は正数でなければなりません(value=$maxlin)" }
        oneByte = ByteArray(1)
        appender = Appender(txtara, maxlin)
    }

    @Synchronized
    fun clear() {
        if (appender != null) {
            appender!!.clear()
        }
    }

    @Synchronized
    override fun close() {
        appender = null
    }

    @Synchronized
    override fun flush() {
        /* empty */
    }

    @Synchronized
    override fun write(`val`: Int) {
        oneByte[0] = `val`.toByte()
        write(oneByte, 0, 1)
    }

    @Synchronized
    override fun write(ba: ByteArray) {
        write(ba, 0, ba.size)
    }

    @Synchronized
    override fun write(ba: ByteArray, str: Int, len: Int) {
        if (appender != null) {
            appender!!.append(bytesToString(ba, str, len))
        }
    }

    internal class Appender(private val textArea: JTextArea, private val maxLines: Int) : Runnable {
        private val lengths: LinkedList<Int> = LinkedList()
        private val values: MutableList<String>
        private var curLength = 0
        private var clear = false
        private var queue = true

        init {
            values = ArrayList()
        }

        @Synchronized
        fun append(`val`: String) {
            values.add(`val`)
            if (queue) {
                queue = false
                EventQueue.invokeLater(this)
            }
        }

        @Synchronized
        fun clear() {
            clear = true
            curLength = 0
            lengths.clear()
            values.clear()
            if (queue) {
                queue = false
                EventQueue.invokeLater(this)
            }
        }

        @Synchronized
        override fun run() {
            if (clear) {
                textArea.text = ""
            }
            values.stream()
                .peek { `val`: String -> curLength += `val`.length }
                .peek { `val`: String ->
                    if (`val`.endsWith(EOL1) || `val`.endsWith(EOL2)) {
                        if (lengths.size >= maxLines) {
                            textArea.replaceRange("", 0, lengths.removeFirst())
                        }
                        lengths.addLast(curLength)
                        curLength = 0
                    }
                }.forEach { str: String? -> textArea.append(str) }
            values.clear()
            clear = false
            queue = true
        }

        companion object {
            private const val EOL1 = "\n"
            private val EOL2 = System.getProperty("line.separator", EOL1)
        }
    }

    companion object {
        private fun bytesToString(ba: ByteArray, str: Int, len: Int): String {
            return try {
                String(ba, str, len, Charset.defaultCharset())
            } catch (thr: UnsupportedEncodingException) {
                String(ba, str, len)
            }
        }
    }
}